tensorflow/core/framework/tensor.cc

```diff
@@ -902,43 +902,31 @@ void Tensor::FillDescription(TensorDescription* description) const {
 }
 
 gtl::InlinedVector<int64, 4> Tensor::ComputeFlatInnerDims(
-    int64 num_out_dims) const {
-  if (num_out_dims == dims()) {
-    return shape_.dim_sizes();
+    gtl::InlinedVector<int64, 4> orig, int64 num_out_dims) {
+  if (num_out_dims == orig.size()) {
+    return orig;
   }
   gtl::InlinedVector<int64, 4> out_dims(num_out_dims, 0);
-  const int64 num_elements = NumElements();
-  int64 prod_out_dims = 1;
-  for (int64 out_dim = num_out_dims - 1; out_dim > 0; --out_dim) {
-    const int64 in_dim = out_dim + (dims() - num_out_dims);
-    out_dims[out_dim] = (in_dim >= dims() || in_dim < 0) ? 1 : dim_size(in_dim);
-    prod_out_dims *= out_dims[out_dim];
-  }
-  if (prod_out_dims != 0) {
-    out_dims[0] = num_elements / prod_out_dims;
-  } else {
-    out_dims[0] = 0;
```

Comment:
Actually, I think there is a bug in the old code that is not discovered by the test case. The old code uses `prod_out_dims` to keep track of the product of the last `num_out_dims` elements and compute the first element of `out_dim` as `num_elements / prod_out_dims`. In the case of `prod_out_dims == 0` it don't make any sense to set the first dimension as `0`.  For example, if you have a tensor with shape `{5,3,0,0,0}` and you run a `flat_inner_dims<?,4>`, you should get a tensor with shape `{15,0,0,0}`. But the way it was computed will gives a `{0,0,0,0}`